connectors/sync_job_runner.py

```diff
@@ -0,0 +1,166 @@
+                )
+                return
+
+            logger.debug(f"Syncing '{self.sync_job.service_type}'")
+            logger.debug(f"Pinging the {self.source_klass} backend")
+            await data_provider.ping()
+            await asyncio.sleep(0)
+
+            mappings = Mappings.default_text_fields_mappings(
+                is_connectors_index=True,
+            )
+
+            logger.debug("Preparing the content index")
+            await self.elastic_server.prepare_content_index(
+                self.sync_job.index_name, mappings=mappings
+            )
+            await asyncio.sleep(0)
+
+            # allows the data provider to change the bulk options
+            bulk_options = self.bulk_options.copy()
+            data_provider.tweak_bulk_options(bulk_options)
+
+            sync_rules_enabled = self.connector.features.sync_rules_enabled()
+
+            if sync_rules_enabled:
+                await validate_filtering(self.connector, self.connector.index)
+
+            result = await self.elastic_server.async_bulk(
+                self.sync_job.index_name,
+                self.prepare_docs(data_provider),
+                self.sync_job.pipeline,
+                filtering=self.sync_job.filtering,
+                sync_rules_enabled=sync_rules_enabled,
+                options=bulk_options,
+            )
+            sync_status = JobStatus.COMPLETED
+            sync_error = None
+        except Exception as e:
+            sync_status = JobStatus.ERROR
+            sync_error = e
+            logger.critical(e, exc_info=True)
+        finally:
+            if sync_status is None:
+                sync_status = JobStatus.ERROR
+            if sync_status == JobStatus.ERROR and sync_error is None:
+                sync_error = "Sync thread didn't finish execution. Check connector logs for more details."
+
+            doc_updated = result.get("doc_updated", 0)
+            doc_created = result.get("doc_created", 0)
+            doc_deleted = result.get("doc_deleted", 0)
+            indexed_count = doc_updated + doc_created
+
+            ingestion_stats = {
+                "indexed_document_count": indexed_count,
+                "indexed_document_volume": 0,
+                "deleted_document_count": 0,
+                "total_document_count": await self.connector.document_count(),
+            }
+            if sync_status == JobStatus.ERROR:
+                await self.sync_job.error(sync_error, ingestion_stats=ingestion_stats)
+            else:
+                await self.sync_job.done(ingestion_stats=ingestion_stats)
+
+            self.sync_job = self.sync_job.reload()
+            await self.connector.sync_done(self.sync_job)
+
+            logger.info(
+                f"Sync done: {indexed_count} indexed, {doc_deleted} "
+                f" deleted. ({int(time.time() - self._start_time)} seconds)"
+            )
+
+    async def _claim_job(self):
+        try:
+            await self.sync_job.claim()
```

Comment:
Hmmm, interesting, so with this PR our connectors will be more multi-tenant friendly?

Reply:
But we don't have concurrency control at all.